File: python/selene_isru/modules/electrolysis.py
from __future__ import annotations

import math
from typing import Any

from ..constants import DEFAULTS, c


OXIDES = [
    ("SiO2", "oxideSiO2", 2, "M_SiO2", "oxideEllinghamSiO2A", "oxideEllinghamSiO2B"),
    ("TiO2", "oxideTiO2", 2, "M_TiO2", "oxideEllinghamTiO2A", "oxideEllinghamTiO2B"),
    ("Al2O3", "oxideAl2O3", 3, "M_Al2O3", "oxideEllinghamAl2O3A", "oxideEllinghamAl2O3B"),
    ("FeO", "oxideFeO", 1, "M_FeO", "oxideEllinghamFeOA", "oxideEllinghamFeOB"),
    ("MgO", "oxideMgO", 1, "M_MgO", "oxideEllinghamMgOA", "oxideEllinghamMgOB"),
    ("CaO", "oxideCaO", 1, "M_CaO", "oxideEllinghamCaOA", "oxideEllinghamCaOB"),
]


def sec_elec_j_per_kg(vcell: float, eta_current: float) -> float:
    return (vcell * 4 * c("F")) / (c("M_O2") * eta_current)


def cp_regolith_j_per_kg_k(t: float) -> float:
    return c("cpRegMaierA") + c("cpRegMaierB") * t + c("cpRegMaierC") / t**2


def sensible_heat_regolith_j_per_kg(tambient: float, tmelt: float, cp_scale: float) -> float:
    a = c("cpRegMaierA")
    b = c("cpRegMaierB")
    c_coef = c("cpRegMaierC")
    base_integral = (
        a * (tmelt - tambient)
        + (b / 2) * (tmelt**2 - tambient**2)
        - c_coef * (1 / tmelt - 1 / tambient)
    )
    return base_integral * (cp_scale / DEFAULTS["cpRegMelt"])


def melt_heat_j_per_kg(params: dict[str, Any]) -> float:
    return sensible_heat_regolith_j_per_kg(params["Tambient"], params["Tmelt"], params["cpRegMelt"]) + params["dHfus"]


def oxide_o2_kg_per_kg(oxygens: int, molar_mass_kg_per_mol: float) -> float:
    return ((oxygens / 2) * c("M_O2")) / molar_mass_kg_per_mol


def oxide_decomposition_voltage(ell_a_j_per_mol_o2: float, ell_b_j_per_mol_o2_k: float, t: float) -> float:
    dgf_j_per_mol_o2 = ell_a_j_per_mol_o2 + ell_b_j_per_mol_o2_k * t
    return -dgf_j_per_mol_o2 / (4 * c("F"))


def mre_voltage_losses(params: dict[str, Any]) -> dict[str, float]:
    j_limit = 4 * c("F") * params["Dox"] * params["Cbulk"] / params["deltaDiff"]
    utilization = min(0.999999, max(0, params["jOperating"] / max(1e-12, j_limit)))
    concentration = c("R") * params["Tmelt"] / (4 * c("F")) * math.log(1 / max(1e-9, 1 - utilization))
    activation = params["mreActivationOverpotentialV"]
    ohmic = params["jOperating"] * params["mreAreaSpecificResistanceOhmM2"]
    return {
        "activationV": activation,
        "ohmicV": ohmic,
        "concentrationV": concentration,
        "availableVoltageV": params["Vcell"] - activation - ohmic - concentration,
        "jLimit_APerM2": j_limit,
    }


def _composition_weighted_voltage(params: dict[str, Any], active: list[bool]) -> float:
    raw_fractions = [max(0, params[oxide[1]]) for oxide in OXIDES]
    scale = max(1, sum(raw_fractions))
    weighted = 0.0
    oxygen_basis = 0.0
    for index, oxide in enumerate(OXIDES):
        if not active[index]:
            continue
        _, _, oxygens, molar_mass_key, ell_a_key, ell_b_key = oxide
        basis = raw_fractions[index] / scale * oxide_o2_kg_per_kg(oxygens, c(molar_mass_key))
        weighted += basis * oxide_decomposition_voltage(c(ell_a_key), c(ell_b_key), params["Tmelt"])
        oxygen_basis += basis
    return weighted / oxygen_basis if oxygen_basis > 0 else 0


def oxide_model_yield(params: dict[str, Any]) -> dict[str, Any]:
    voltage = mre_voltage_losses(params)
    if not params["oxideModel"]:
        return {
            "xO2Effective": params["xO2"] * params["fExtract"],
            "oxideYield": [{"oxide": oxide[0], "o2KgPerKg": 0, "decomposed": False} for oxide in OXIDES],
            "reversibleVoltageV": _composition_weighted_voltage(params, [True for _ in OXIDES]),
            "availableVoltageV": voltage["availableVoltageV"],
        }

    raw_fractions = [max(0, params[oxide[1]]) for oxide in OXIDES]
    raw_total = sum(raw_fractions)
    if raw_total <= 0:
        return {
            "xO2Effective": params["xO2"] * params["fExtract"],
            "oxideYield": [{"oxide": oxide[0], "o2KgPerKg": 0, "decomposed": False} for oxide in OXIDES],
            "reversibleVoltageV": 0,
            "availableVoltageV": voltage["availableVoltageV"],
        }

    fraction_scale = max(1, raw_total)
    recovery = params["fExtract"] * c("oxideRecoveryCalibration")
    x_o2_effective = 0.0
    oxide_yield: list[dict[str, Any]] = []

    for i, oxide in enumerate(OXIDES):
        oxide_name, _, oxygens, molar_mass_key, ell_a_key, ell_b_key = oxide
        mass_frac = raw_fractions[i] / fraction_scale
        decomposed = oxide_decomposition_voltage(c(ell_a_key), c(ell_b_key), params["Tmelt"]) <= voltage["availableVoltageV"]
        o2_kg_per_kg = (
            mass_frac * oxide_o2_kg_per_kg(oxygens, c(molar_mass_key)) * recovery
            if decomposed
            else 0
        )
        x_o2_effective += o2_kg_per_kg
        oxide_yield.append(
            {
                "oxide": oxide_name,
                "o2KgPerKg": o2_kg_per_kg,
                "decomposed": decomposed,
            }
        )

    return {
        "xO2Effective": max(1e-9, x_o2_effective),
        "oxideYield": oxide_yield,
        "reversibleVoltageV": _composition_weighted_voltage(params, [row["decomposed"] for row in oxide_yield]),
        "availableVoltageV": voltage["availableVoltageV"],
    }


def simulate_electrolysis(params: dict[str, Any]) -> dict[str, Any]:
    sec_elec_j_per_kg_value = sec_elec_j_per_kg(params["Vcell"], params["etaCurrent"])
    voltage_losses = mre_voltage_losses(params)
    oxide_model = oxide_model_yield(params)
    r_reg = 1 / oxide_model["xO2Effective"]
    q_melt = melt_heat_j_per_kg(params)
    sec_thermal_j_per_kg = r_reg * q_melt
    sec_parasitic_j_per_kg = params["fParasitic"] * (sec_elec_j_per_kg_value + sec_thermal_j_per_kg)
    mdot_o2_kg_per_s = params["targetKgPerDay"] / 86400
    current_a = mdot_o2_kg_per_s * 4 * c("F") / (c("M_O2") * params["etaCurrent"])
    melt_viscosity_pa_s = params["Amu"] * params["Tmelt"] * math.exp(params["Bmu"] / (params["Tmelt"] - params["T0vft"]))
    drain_velocity_m_per_s = (
        params["rhoSlag"] * c("gL") * params["hMelt"] ** 2 * math.sin(params["thetaDrain"])
    ) / (3 * melt_viscosity_pa_s)
    j_limit_a_per_m2 = voltage_losses["jLimit_APerM2"]
    j_operating_a_per_m2 = params["jOperating"]
    electrode_area_m2 = current_a / j_operating_a_per_m2
    current_utilization = j_operating_a_per_m2 / j_limit_a_per_m2
    modeled_required_voltage = oxide_model["reversibleVoltageV"] + voltage_losses["activationV"] + voltage_losses["ohmicV"] + voltage_losses["concentrationV"]
    voltage_margin = params["Vcell"] - modeled_required_voltage
    unallocated_voltage = max(0, voltage_margin)
    electrical_input_w = current_a * params["Vcell"]
    chemical_power_w = current_a * params["etaCurrent"] * oxide_model["reversibleVoltageV"]
    modeled_loss_power_w = max(0, electrical_input_w - chemical_power_w)
    reactor_mass_kg = params["kReactorMass"] * params["targetKgPerDay"]
    warnings: list[dict[str, Any]] = []

    if j_operating_a_per_m2 > 0.85 * j_limit_a_per_m2:
        warnings.append(
            {
                "id": "anode-current",
                "severity": "alarm",
                "module": "electrolysis",
                "message": "Operating current density exceeds 85% of limiting current density.",
                "value": j_operating_a_per_m2,
                "limit": 0.85 * j_limit_a_per_m2,
            }
        )
    if voltage_margin < 0:
        warnings.append({"id": "mre-voltage-shortfall", "severity": "alarm", "module": "electrolysis", "message": "Applied MRE voltage is below the modeled reversible, activation, ohmic, and concentration requirement.", "value": params["Vcell"], "limit": modeled_required_voltage})
    if params["oxideModel"] and all(not row["decomposed"] for row in oxide_model["oxideYield"]):
        minimum_voltage = min(oxide_decomposition_voltage(c(oxide[4]), c(oxide[5]), params["Tmelt"]) for oxide in OXIDES)
        warnings.append({"id": "mre-no-oxide-yield", "severity": "alarm", "module": "electrolysis", "message": "No modeled oxide is energetically accessible at the current voltage-loss operating point.", "value": oxide_model["availableVoltageV"], "limit": minimum_voltage})

    return {
        "secElec_JPerKg": sec_elec_j_per_kg_value,
        "secThermal_JPerKg": sec_thermal_j_per_kg,
        "secParasitic_JPerKg": sec_parasitic_j_per_kg,
        "currentA": current_a,
        "cellVoltageV": params["Vcell"],
        "jLimit_APerM2": j_limit_a_per_m2,
        "jOperating_APerM2": j_operating_a_per_m2,
        "meltViscosityPaS": melt_viscosity_pa_s,
        "drainVelocityMPerS": drain_velocity_m_per_s,
        "reactorMassKg": reactor_mass_kg,
        "xO2Effective": oxide_model["xO2Effective"],
        "oxideYield": oxide_model["oxideYield"],
        "reversibleVoltageV": oxide_model["reversibleVoltageV"],
        "activationOverpotentialV": voltage_losses["activationV"],
        "ohmicOverpotentialV": voltage_losses["ohmicV"],
        "concentrationOverpotentialV": voltage_losses["concentrationV"],
        "unallocatedVoltageV": unallocated_voltage,
        "voltageMarginV": voltage_margin,
        "electrodeAreaM2": electrode_area_m2,
        "currentUtilization": current_utilization,
        "electricalInputW": electrical_input_w,
        "chemicalPowerW": chemical_power_w,
        "modeledLossPowerW": modeled_loss_power_w,
        "warnings": warnings,
    }

File: packages/engine/src/modules/electrolysis.ts
import { DEFAULTS, PHYSICAL_CONSTANTS } from "../constants";
import type { OxideYield, SimParams, Warning } from "../types";

export interface ElectrolysisOutput {
  secElec_JPerKg: number;
  secThermal_JPerKg: number;
  secParasitic_JPerKg: number;
  currentA: number;
  cellVoltageV: number;
  jLimit_APerM2: number;
  jOperating_APerM2: number;
  meltViscosityPaS: number;
  drainVelocityMPerS: number;
  reactorMassKg: number;
  xO2Effective: number;
  oxideYield: OxideYield[];
  reversibleVoltageV: number;
  activationOverpotentialV: number;
  ohmicOverpotentialV: number;
  concentrationOverpotentialV: number;
  unallocatedVoltageV: number;
  voltageMarginV: number;
  electrodeAreaM2: number;
  currentUtilization: number;
  electricalInputW: number;
  chemicalPowerW: number;
  modeledLossPowerW: number;
  warnings: Warning[];
}

interface OxideDef {
  oxide: string;
  fraction: (params: SimParams) => number;
  oxygens: number;
  molarMassKgPerMol: number;
  ellA_JPerMolO2: number;
  ellB_JPerMolO2K: number;
}

interface OxideModelOutput {
  xO2Effective: number;
  oxideYield: OxideYield[];
  reversibleVoltageV: number;
  availableVoltageV: number;
}

interface VoltageLosses {
  activationV: number;
  ohmicV: number;
  concentrationV: number;
  availableVoltageV: number;
  jLimit_APerM2: number;
}

const OXIDES: OxideDef[] = [
  {
    oxide: "SiO2",
    fraction: (params) => params.oxideSiO2,
    oxygens: 2,
    molarMassKgPerMol: PHYSICAL_CONSTANTS.M_SiO2.value,
    ellA_JPerMolO2: PHYSICAL_CONSTANTS.oxideEllinghamSiO2A.value,
    ellB_JPerMolO2K: PHYSICAL_CONSTANTS.oxideEllinghamSiO2B.value
  },
  {
    oxide: "TiO2",
    fraction: (params) => params.oxideTiO2,
    oxygens: 2,
    molarMassKgPerMol: PHYSICAL_CONSTANTS.M_TiO2.value,
    ellA_JPerMolO2: PHYSICAL_CONSTANTS.oxideEllinghamTiO2A.value,
    ellB_JPerMolO2K: PHYSICAL_CONSTANTS.oxideEllinghamTiO2B.value
  },
  {
    oxide: "Al2O3",
    fraction: (params) => params.oxideAl2O3,
    oxygens: 3,
    molarMassKgPerMol: PHYSICAL_CONSTANTS.M_Al2O3.value,
    ellA_JPerMolO2: PHYSICAL_CONSTANTS.oxideEllinghamAl2O3A.value,
    ellB_JPerMolO2K: PHYSICAL_CONSTANTS.oxideEllinghamAl2O3B.value
  },
  {
    oxide: "FeO",
    fraction: (params) => params.oxideFeO,
    oxygens: 1,
    molarMassKgPerMol: PHYSICAL_CONSTANTS.M_FeO.value,
    ellA_JPerMolO2: PHYSICAL_CONSTANTS.oxideEllinghamFeOA.value,
    ellB_JPerMolO2K: PHYSICAL_CONSTANTS.oxideEllinghamFeOB.value
  },
  {
    oxide: "MgO",
    fraction: (params) => params.oxideMgO,
    oxygens: 1,
    molarMassKgPerMol: PHYSICAL_CONSTANTS.M_MgO.value,
    ellA_JPerMolO2: PHYSICAL_CONSTANTS.oxideEllinghamMgOA.value,
    ellB_JPerMolO2K: PHYSICAL_CONSTANTS.oxideEllinghamMgOB.value
  },
  {
    oxide: "CaO",
    fraction: (params) => params.oxideCaO,
    oxygens: 1,
    molarMassKgPerMol: PHYSICAL_CONSTANTS.M_CaO.value,
    ellA_JPerMolO2: PHYSICAL_CONSTANTS.oxideEllinghamCaOA.value,
    ellB_JPerMolO2K: PHYSICAL_CONSTANTS.oxideEllinghamCaOB.value
  }
];

export function secElecJPerKg(Vcell: number, etaCurrent: number): number {
  return (Vcell * 4 * PHYSICAL_CONSTANTS.F.value) / (PHYSICAL_CONSTANTS.M_O2.value * etaCurrent);
}

export function cpRegolithJPerKgK(T: number): number {
  return (
    PHYSICAL_CONSTANTS.cpRegMaierA.value +
    PHYSICAL_CONSTANTS.cpRegMaierB.value * T +
    PHYSICAL_CONSTANTS.cpRegMaierC.value / T ** 2
  );
}

export function sensibleHeatRegolithJPerKg(Tambient: number, Tmelt: number, cpScale: number): number {
  const a = PHYSICAL_CONSTANTS.cpRegMaierA.value;
  const b = PHYSICAL_CONSTANTS.cpRegMaierB.value;
  const c = PHYSICAL_CONSTANTS.cpRegMaierC.value;
  const baseIntegral =
    a * (Tmelt - Tambient) +
    (b / 2) * (Tmelt ** 2 - Tambient ** 2) -
    c * (1 / Tmelt - 1 / Tambient);
  return baseIntegral * (cpScale / DEFAULTS.cpRegMelt);
}

export function meltHeatJPerKg(params: SimParams): number {
  return sensibleHeatRegolithJPerKg(params.Tambient, params.Tmelt, params.cpRegMelt) + params.dHfus;
}

export function oxideO2KgPerKg(oxygens: number, molarMassKgPerMol: number): number {
  return ((oxygens / 2) * PHYSICAL_CONSTANTS.M_O2.value) / molarMassKgPerMol;
}

export function oxideDecompositionVoltage(ellA_JPerMolO2: number, ellB_JPerMolO2K: number, T: number): number {
  const dGf_JPerMolO2 = ellA_JPerMolO2 + ellB_JPerMolO2K * T;
  return -dGf_JPerMolO2 / (4 * PHYSICAL_CONSTANTS.F.value);
}

export function mreVoltageLosses(params: SimParams): VoltageLosses {
  const jLimit_APerM2 =
    (4 * PHYSICAL_CONSTANTS.F.value * params.Dox * params.Cbulk) / params.deltaDiff;
  const utilization = Math.min(0.999999, Math.max(0, params.jOperating / Math.max(1e-12, jLimit_APerM2)));
  const concentrationV =
    (PHYSICAL_CONSTANTS.R.value * params.Tmelt / (4 * PHYSICAL_CONSTANTS.F.value)) *
    Math.log(1 / Math.max(1e-9, 1 - utilization));
  const activationV = params.mreActivationOverpotentialV;
  const ohmicV = params.jOperating * params.mreAreaSpecificResistanceOhmM2;
  return {
    activationV,
    ohmicV,
    concentrationV,
    availableVoltageV: params.Vcell - activationV - ohmicV - concentrationV,
    jLimit_APerM2
  };
}

function compositionWeightedVoltage(params: SimParams, active: boolean[]): number {
  const rawFractions = OXIDES.map((oxide) => Math.max(0, oxide.fraction(params)));
  const scale = Math.max(1, rawFractions.reduce((total, value) => total + value, 0));
  let weighted = 0;
  let oxygenBasis = 0;
  for (let index = 0; index < OXIDES.length; index += 1) {
    if (!active[index]) continue;
    const oxide = OXIDES[index]!;
    const basis = rawFractions[index]! / scale * oxideO2KgPerKg(oxide.oxygens, oxide.molarMassKgPerMol);
    weighted += basis * oxideDecompositionVoltage(oxide.ellA_JPerMolO2, oxide.ellB_JPerMolO2K, params.Tmelt);
    oxygenBasis += basis;
  }
  return oxygenBasis > 0 ? weighted / oxygenBasis : 0;
}

export function oxideModelYield(params: SimParams): OxideModelOutput {
  const voltage = mreVoltageLosses(params);
  if (!params.oxideModel) {
    return {
      xO2Effective: params.xO2 * params.fExtract,
      oxideYield: OXIDES.map((oxide) => ({
        oxide: oxide.oxide,
        o2KgPerKg: 0,
        decomposed: false
      })),
      reversibleVoltageV: compositionWeightedVoltage(params, OXIDES.map(() => true)),
      availableVoltageV: voltage.availableVoltageV
    };
  }

  const rawFractions = OXIDES.map((oxide) => Math.max(0, oxide.fraction(params)));
  const rawTotal = rawFractions.reduce((total, value) => total + value, 0);
  if (rawTotal <= 0) {
    return {
      xO2Effective: params.xO2 * params.fExtract,
      oxideYield: OXIDES.map((oxide) => ({
        oxide: oxide.oxide,
        o2KgPerKg: 0,
        decomposed: false
      })),
      reversibleVoltageV: 0,
      availableVoltageV: voltage.availableVoltageV
    };
  }

  const fractionScale = Math.max(1, rawTotal);
  const recovery = params.fExtract * PHYSICAL_CONSTANTS.oxideRecoveryCalibration.value;
  let xO2Effective = 0;
  const oxideYield: OxideYield[] = [];

  for (let i = 0; i < OXIDES.length; i += 1) {
    const oxide = OXIDES[i]!;
    const massFrac = rawFractions[i]! / fractionScale;
    const decomposed =
      oxideDecompositionVoltage(oxide.ellA_JPerMolO2, oxide.ellB_JPerMolO2K, params.Tmelt) <=
      voltage.availableVoltageV;
    const o2KgPerKg = decomposed ? massFrac * oxideO2KgPerKg(oxide.oxygens, oxide.molarMassKgPerMol) * recovery : 0;
    xO2Effective += o2KgPerKg;
    oxideYield.push({
      oxide: oxide.oxide,
      o2KgPerKg,
      decomposed
    });
  }

  return {
    xO2Effective: Math.max(1e-9, xO2Effective),
    oxideYield,
    reversibleVoltageV: compositionWeightedVoltage(params, oxideYield.map((row) => row.decomposed)),
    availableVoltageV: voltage.availableVoltageV
  };
}

export function simulateElectrolysis(params: SimParams): ElectrolysisOutput {
  const secElec_JPerKg = secElecJPerKg(params.Vcell, params.etaCurrent);
  const voltageLosses = mreVoltageLosses(params);
  const oxideModel = oxideModelYield(params);
  const Rreg = 1 / oxideModel.xO2Effective;
  const Qmelt = meltHeatJPerKg(params);
  const secThermal_JPerKg = Rreg * Qmelt;
  const secParasitic_JPerKg = params.fParasitic * (secElec_JPerKg + secThermal_JPerKg);
  const mdotO2_kgPerS = params.targetKgPerDay / 86400;
  const currentA =
    (mdotO2_kgPerS * 4 * PHYSICAL_CONSTANTS.F.value) /
    (PHYSICAL_CONSTANTS.M_O2.value * params.etaCurrent);
  const meltViscosityPaS = params.Amu * params.Tmelt * Math.exp(params.Bmu / (params.Tmelt - params.T0vft));
  const drainVelocityMPerS =
    (params.rhoSlag * PHYSICAL_CONSTANTS.gL.value * params.hMelt ** 2 * Math.sin(params.thetaDrain)) /
    (3 * meltViscosityPaS);
  const jLimit_APerM2 = voltageLosses.jLimit_APerM2;
  const jOperating_APerM2 = params.jOperating;
  const electrodeAreaM2 = currentA / jOperating_APerM2;
  const currentUtilization = jOperating_APerM2 / jLimit_APerM2;
  const modeledRequiredVoltageV =
    oxideModel.reversibleVoltageV + voltageLosses.activationV + voltageLosses.ohmicV + voltageLosses.concentrationV;
  const voltageMarginV = params.Vcell - modeledRequiredVoltageV;
  const unallocatedVoltageV = Math.max(0, voltageMarginV);
  const electricalInputW = currentA * params.Vcell;
  const chemicalPowerW = currentA * params.etaCurrent * oxideModel.reversibleVoltageV;
  const modeledLossPowerW = Math.max(0, electricalInputW - chemicalPowerW);
  const reactorMassKg = params.kReactorMass * params.targetKgPerDay;
  const warnings: Warning[] = [];

  if (jOperating_APerM2 > 0.85 * jLimit_APerM2) {
    warnings.push({
      id: "anode-current",
      severity: "alarm",
      module: "electrolysis",
      message: "Operating current density exceeds 85% of limiting current density.",
      value: jOperating_APerM2,
      limit: 0.85 * jLimit_APerM2
    });
  }
  if (voltageMarginV < 0) {
    warnings.push({
      id: "mre-voltage-shortfall",
      severity: "alarm",
      module: "electrolysis",
      message: "Applied MRE voltage is below the modeled reversible, activation, ohmic, and concentration requirement.",
      value: params.Vcell,
      limit: modeledRequiredVoltageV
    });
  }
  if (oxideModel.oxideYield.every((row) => !row.decomposed) && params.oxideModel) {
    warnings.push({
      id: "mre-no-oxide-yield",
      severity: "alarm",
      module: "electrolysis",
      message: "No modeled oxide is energetically accessible at the current voltage-loss operating point.",
      value: oxideModel.availableVoltageV,
      limit: Math.min(...OXIDES.map((oxide) => oxideDecompositionVoltage(oxide.ellA_JPerMolO2, oxide.ellB_JPerMolO2K, params.Tmelt)))
    });
  }

  return {
    secElec_JPerKg,
    secThermal_JPerKg,
    secParasitic_JPerKg,
    currentA,
    cellVoltageV: params.Vcell,
    jLimit_APerM2,
    jOperating_APerM2,
    meltViscosityPaS,
    drainVelocityMPerS,
    reactorMassKg,
    xO2Effective: oxideModel.xO2Effective,
    oxideYield: oxideModel.oxideYield,
    reversibleVoltageV: oxideModel.reversibleVoltageV,
    activationOverpotentialV: voltageLosses.activationV,
    ohmicOverpotentialV: voltageLosses.ohmicV,
    concentrationOverpotentialV: voltageLosses.concentrationV,
    unallocatedVoltageV,
    voltageMarginV,
    electrodeAreaM2,
    currentUtilization,
    electricalInputW,
    chemicalPowerW,
    modeledLossPowerW,
    warnings
  };
}
